### PR TITLE
feat: add MiniCPM4.1 SparDA forecast selection

### PR DESCRIPTION
**Pull request by Hao-tian-Zheng**
## Motivation

This PR adds the first, deliberately scoped SGLang integration point for SparDA: a reusable cross-layer Forecast selector runtime, with MiniCPM4.1 as the first model adapter.

The feature is opt-in and disabled by default. It does not include host KV-cache offload or asynchronous prefetch, so this PR does not claim to be the complete SparDA runtime or to provide an inference-speedup result.

## Modifications

- Add a `ForecastState` owned by each `ForwardBatch` for per-request/per-layer forecast handoff, including first-layer current-query fallback, reset, and batch isolation.
- Add `--enable-sparda` and `--sparda-indexer-path`, with startup validation for model type, pipeline parallelism, path existence, checkpoint keys, and tensor shapes.
- Add the MiniCPM4.1 (`model_type=minicpm`) adapter with the official `q_future_proj` layout, RoPE handling, GQA/KV-head slicing, and TP-compatible loading.
- Map the official indexer checkpoint layout, including nested state-dict formats and `module.` prefixes. The official `q_curr_proj` entry is accepted as an extra checkpoint key; PR1 intentionally uses the current-query fallback for the first layer.
- Thread the optional forecast selector through `RadixAttention` and the MiniCPM sparse-attention backend. The existing current-query selector, DSA, MiniCPM-SALA, and default attention paths remain unchanged when the flag is disabled.
- Keep the forecast selector on the reference compressed-attention selection path; no host KV residency, copy stream, event synchronization, eviction, or prefetch logic is included.
- Add documentation and registered unit tests. No model checkpoint or training data is committed.

## Accuracy Tests

Targeted registered tests were run in a remote NVIDIA A100 validation environment:

```text
77 passed, 17 warnings, 8 subtests passed in 19.73s
```

The command covered:

```text
test/registered/unit/layers/test_lookahead_forecast_state.py
test/registered/unit/layers/test_minicpm_sparse_metadata.py
test/registered/unit/configs/test_minicpm_config.py
test/registered/unit/models/test_minicpm_sparda.py
test/registered/unit/layers/test_minicpm_attention_adapter.py
test/registered/unit/layers/test_radix_attention.py
test/registered/unit/layers/attention/test_dsa_mqa_logits_chunking.py
```

Additional validation passed:

- Synthetic no-tie selector correctness: the runtime top-1 block index matched the independent reference block reduction.
- MiniCPM4.1 greedy generation smoke: single-request runs completed 16 output tokens, and four-request runs completed 64 aggregate output tokens.
- TP1 and TP2 greedy generation smoke both succeeded, concurrent requests returned independent results, and a no-flag MiniCPM4.1 startup/warmup succeeded through the default path.

The official trained Forecast checkpoint was not available from the released SparDA artifacts, so official top-k overlap/recall and end-to-end quality comparisons are not reported here. The synthetic checkpoint is used only to validate loading and runtime wiring.

## Speed Tests and Profiling

The following are warm-run measurements with deterministic inputs and 16 generated tokens per request. The 32K and 64K cases use 32768 and 65504 input tokens respectively, so the latter stays within the 65536-token context limit. `TTFT / decode tok/s` is reported in seconds and aggregate generated tokens per second.

| Context | Batch | Default path TTFT / decode tok/s | `--enable-sparda` TTFT / decode tok/s | Output tokens |
| --- | ---: | ---: | ---: | ---: |
| 32K | 1 | 4.08 / 33.21 | 5.62 / 15.89 | 16 |
| 64K | 1 | 11.12 / 34.72 | 11.68 / 15.12 | 16 |
| 32K | 4 | 19.96 / 138.04 | 23.00 / 63.49 | 64 |
| 64K | 4 | 45.04 / 138.87 | 47.82 / 60.89 | 64 |

The default column is the normal non-SparDA dense path. The enabled column uses MiniCPM sparse defaults and a synthetic Forecast checkpoint. These are selection-only measurements: there is no host KV prefetch/offload, so no end-to-end SparDA speedup is claimed. GPU-memory and isolated selector-overhead measurements are omitted because the validation environment is shared; they will be measured with the independent KV-prefetch PR.

## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit). All applicable changed-file pre-commit hooks passed, including formatting, lint, spelling, and registered-test checks. This PR has no Rust files, so Rust hooks had no files to check.
- [x] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [x] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [x] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed). Synthetic selector correctness, greedy-generation smoke tests, and 32K/64K batch-1/4 selection-only measurements are reported above; official-checkpoint quality and host-KV-prefetch measurements are outside this PR's scope or unavailable.
- [x] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review and Merge Process

1. Ping Merge Oncalls to start the process. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
2. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
3. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
   - Common commands include `/tag-and-rerun-ci`, `/tag-run-ci-label`, `/rerun-failed-ci`
4. After green CI and required approvals, ask Merge Oncalls or people with Write permission to merge the PR.

---
### CI States

Latest PR Test (Base): :hourglass_flowing_sand: [Run #34142024450](https://github.com/sgl-project/sglang/actions/runs/34142024450)
Latest PR Test (Extra): :hourglass_flowing_sand: [Run #34142024315](https://github.com/sgl-project/sglang/actions/runs/34142024315)
Latest PR Test (AMD ROCm 7.2): :hourglass_flowing_sand: [Run #34142024459](https://github.com/sgl-project/sglang/actions/runs/34142024459)

### CODE REVIEW

**Comment by Hao-tian-Zheng** [comment]
@Fridge003 @merrymercy @Qiaolin-Yu Could you please add the `run-ci` and `run-ci-extra` labels or trigger CI for this PR? The author slash command can rerun workflows but does not have permission to add those labels. PR1 is intentionally limited to MiniCPM4.1 Forecast selection; host KV prefetch/offload is deferred to Phase 2. Targeted tests passed in a remote NVIDIA A100 validation environment (75 passed).
